en/tutorials/chat-app.md

```diff
@@ -0,0 +1,400 @@
+  var fileInput = $('<input type="file" accept="image/png, image/jpeg, image/gif" data-action="uploadAvatar" />');
+
+  // create a container for that input to cover the hint message
+  var inputContainer = $('<div class="bg-white input-container"></div>');
+
+  // set a change listener for triggering the img upload process before attaching in the input to the DOM
+  fileInput.on('change', handleImgUpload);
+
+  // put the file input into the container and attach it to the DOM
+  inputContainer.append(fileInput);
+  parent.prepend(inputContainer);
+}
+```
+
+That ^ seems simple enough, right? Next is what happens once someone selects an img from the filepicker input:
+
+```js 
+function handleImgUpload(e) {
+  // process the image for saving to the avatarStore with the file taken from the file input's change event
+  avatarProcess(e.target.files[0]);
+
+  then remove the input from the DOM
+  e.target.remove();
+}
+
+function avatarProcess(imgData) {
+  // define a function to model the avatar data before sending it off to be saved
+  function handleImg(img) {
+    return function process(e) {
+      var src = e.target.result;
+      var props = {
+        src: src,
+        id: hoodie.account.username
+      };
+
+      saveAvatar(props);
+    };
+  }
+
+  // create a new FileReader, set our pre-defined function to trigger once the file reader has our image loaded, and trigger the read event
+  var reader = new FileReader();
+  reader.onload = handleImg(imgData);
+  reader.readAsDataURL(imgData);
+}
+
+function saveAvatar(props) {
+  // try to update a user's avatar, otherwise create a new item in the store for a user's avatar data, then update their image in the chat stream
+  hoodie.store.updateOrAdd('avatar', hoodie.account.username, props).publish().then(function(avatar) {
+    var avatarEl = $('[data-avatar="'+avatar[0].id+'"]');
+    avatarEl[0].src = avatar[0].src;
+  })
+  .catch(function(error) { console.log(error); })
+}
+```
+
+Finally, we can define a function that fetches a user's avatar when a new message is added or when a user signs into the app:
+
+```js
+function fetchAvatar(user) {
+  var user = user || hoodie.account.username;
+  var imgEl = $('[data-avatar="'+user+'"]');
+  
+  hoodie.global.find('avatar', user)
+    .then(function(avatar) { imgEl.attr('src', avatar.src); })
+    .catch(function(error) { console.log(user, error); return; });
+}
+
+// replace currentUser avatar data with the current username, if there is one, then fetch the real avatar.
+if ( hoodie.account.username ) {
+  $('[data-avatar="currentUser"]').attr('data-avatar', hoodie.account.username);
+  fetchAvatar();
+}
+```
+
+This is definitely one of the more complex feature of our app but worth it to make this chat app feel more complete. As an extra challenge, try to add notifications for when the current user is @mentioned by another user in the chat by highlighting their messages with a unique background color. 
```

Comment:
I think this might read better as "more complex features" rather than "more complex feature".

Hope this helps :)